src/graphics/Window.zig

```diff
@@ -10,16 +12,283 @@ pub const c = @cImport ({
+									releaseAction();
+								}
+							}
+						}
+					}
+				} else {
+					const axis = key.gamepadAxis.?.axis;
+					const positive = key.gamepadAxis.?.positive;
+					var newAxis = applyDeadzone(newState.axes[@intCast(axis)]);
+					var oldAxis = applyDeadzone(oldState.axes[@intCast(axis)]);
+					if(!positive) {
+						newAxis *= -1.0;
+						oldAxis *= -1.0;
+					}
+					newAxis = @max(newAxis, 0.0);
+					oldAxis = @max(oldAxis, 0.0);
+					const oldPressed = oldAxis > 0.5;
+					const newPressed = newAxis > 0.5;
+					if (oldPressed != newPressed) {
+						key.pressed = newPressed;
+						if (newPressed) {
+							if (key.pressAction) |pressAction| {
+								pressAction();
+							}
+						} else {
+							if (key.releaseAction) |releaseAction| {
+								releaseAction();
+							}
+						}
+					}
+					if (newAxis != oldAxis) {
+						key.value = newAxis;
+					}
+				}
+			}
+		}
+		if (!grabbed) {
+			const x = main.KeyBoard.key("uiRight").value - main.KeyBoard.key("uiLeft").value;
+			const y = main.KeyBoard.key("uiDown").value - main.KeyBoard.key("uiUp").value;
+			if (x != 0 or y != 0) {
+				lastUsedMouse = false;
+				GLFWCallbacks.currentPos[0] += @floatCast(x * delta * 256);
+				GLFWCallbacks.currentPos[1] += @floatCast(y * delta * 256);
+				const winSize = getWindowSize();
+				GLFWCallbacks.currentPos[0] = std.math.clamp(GLFWCallbacks.currentPos[0], 0, winSize[0]);
+				GLFWCallbacks.currentPos[1] = std.math.clamp(GLFWCallbacks.currentPos[1], 0, winSize[1]);
+			}
+		}
+		scrollOffset += @floatCast((main.KeyBoard.key("scrollUp").value - main.KeyBoard.key("scrollDown").value) * delta * 4);
+		setCursorVisible(!grabbed and lastUsedMouse);
+	}
+	pub fn isControllerConnected() bool {
+		return gamepadState.count() > 0;
+	}
+	pub fn wereControllerMappingsDownloaded() bool {
+		return controllerMappingsDownloaded.load(std.builtin.AtomicOrder.acquire);
+	}
+	const ControllerMappingDownloadTask = struct { // MARK: ControllerMappingDownloadTask
+		curTimestamp: i128,
+		var running = std.atomic.Value(bool).init(false);
+		const vtable = main.utils.ThreadPool.VTable{
+			.getPriority = @ptrCast(&getPriority),
+			.isStillNeeded = @ptrCast(&isStillNeeded),
+			.run = @ptrCast(&run),
+			.clean = @ptrCast(&clean),
+		};
+
+		pub fn schedule(curTimestamp: i128) void {
+
+			if (running.swap(true, .monotonic)) {
+				std.log.warn("Attempt to schedule a duplicate controller mapping download task!", .{});
+				return; // Controller mappings are already downloading.
+			}
+			controllerMappingsDownloaded.store(false, .monotonic);
+			const task = main.globalAllocator.create(ControllerMappingDownloadTask);
+			task.* = ControllerMappingDownloadTask {
+				.curTimestamp = curTimestamp,
+			};
+			main.threadPool.addTask(task, &vtable);
+		}
+
+		pub fn getPriority(_: *ControllerMappingDownloadTask) f32 {
+			return std.math.inf(f32);
+		}
+
+		pub fn isStillNeeded(_: *ControllerMappingDownloadTask, _: i64) bool {
```

Comment:
The second parameter was just removed in e863e9f311a785f8c98eeb35a0e664de8512dd2e